### DISCUSSION TITLE
Performance issues

### DISCUSSION THREAD

**Discussion by jeffdlange** [maximgurin/verifica#5] [General]
Greetings, and first off thank you @maximgurin for the verifica gem 👏

I was drawn to use your work because it's great combination of flexibility and simplicity--or, in a word, elegance. I was also excited that you'd already tested it in production envs with good performance outcomes.

### The Question
Which brings us to my recent issues. I'm getting much worse performance than expected in my application. To summarize the problem:
- I have a Students table (with about 10k rows)
- I have potentially (and commonly) hundreds of elements in each row's `read_allow_sids` column
- `current_user`'s sids may also commonly be hundreds of elements long
- **MAIN ISSUE**: My auth query is slow; about 4.5 seconds long to execute (i.e. the query recommended by verifica's docs: `where(read_allow_sids && ARRAY[...]).not(read_deny_sids && ARRAY[...])`).
- I added a GIN index to `read_allow_sids` and `read_deny_sids`. This makes the query execute in 200ms. Nice, that would probably be good enough.
- However...the catch is that by default postgres is ignoring my GIN index. The only way I've gotten it to work is to force it by setting `SET enable_seqscan TO OFF`, which is obviously not actually a viable option. 

### Conclusion
Have you encountered any performance issues like this? How did you overcome them? Do you know how to get postgres to respect the GIN index?

Thanks for your time, and all the work you put into making this gem!

For reference, this question is also described in [this stack overflow question](https://stackoverflow.com/questions/76883801/postgres-is-ignoring-my-gin-index-on-an-array-strings-column).

**Comment by maximgurin** [comment]
Hi Jeff,

Thank you so much for the positive feedback on Verifica 🙏
It's gratifying as an author to know that others find it valuable.

I'm not well-versed in the GIN index and Postgres nuances, but I can shed light on how we employed Verifica in 2 similar projects I recently worked on. In both, Elasticsearch was central to the architecture. Beyond just authorization queries, it powered user-driven free-text searches and faceted searches where authorization was seamlessly integrated.

To illustrate, let's look at your Students table:
1. Retain the Students in Postgres as the primary data source.
2. Create an Elasticsearch index for Students and establish a synchronization mechanism from Postgres to Elasticsearch. While there are automated methods for this, manual approaches (like callback or event-driven ones) can also be effective.
3. Execute the authorization queries in Elasticsearch rather than Postgres.

One of the standout features of this method is the ability to incorporate user-generated search queries while ensuring only authorized results are shown. For instance, a user might search for Students with the name "John*", but will only see the authorized entries. Something like this:
```json
{
  "filtered": {
    "query": {
      "query_string": {
        "query": "John*"
      }
    },
    "filter": {
      "bool": {
        "should": [{
          "terms": {
            "read_allow_sids": ["sid_1", "sid_2", "etc."]
          }
        }],
        "must_not": [{
          "terms": {
            "read_deny_sids": ["sid_1", "sid_2", "etc."]
          }
        }]
      }
    }
  }
}
```

Certainly, you could achieve something similar using just Postgres. However, in my experience, Elasticsearch offered better performance for these queries and was more straightforward to handle. Of course, results might differ based on your specific circumstances.

Hope this helps!

**Reply by jeffdlange** [reply]
I haven't used Elastic Search before, but I will do some research on it and see if this fits my use case. Thank you for your timely response!
